File: Leetcode/leetcode-0145-BinaryTreePostorderTraversal-[tree]-[hard]/Solution.iterative.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
class Solution {
    public List<Integer> postorderTraversal(TreeNode root) {
        ArrayList<Integer> result = new ArrayList<>();
        if (root == null) {
            return result;
        }
        
        Stack<TreeNode> stack = new Stack<>();
        stack.push(root);
        while (!stack.isEmpty()) {
            TreeNode cur = stack.pop();
            result.add(0, cur.val);
            
            if (cur.left != null) {
                stack.push(cur.left);
            }
            
            if (cur.right != null) {
                stack.push(cur.right);
            }
        }
        
        return result;
    }
}

File: Leetcode/leetcode-0145-BinaryTreePostorderTraversal-[tree]-[hard]/Solution.iterative.js
/**
 * Definition for a binary tree node.
 * function TreeNode(val, left, right) {
 *     this.val = (val===undefined ? 0 : val)
 *     this.left = (left===undefined ? null : left)
 *     this.right = (right===undefined ? null : right)
 * }
 */
/**
 * @param {TreeNode} root
 * @return {number[]}
 */
var postorderTraversal = function (root) {
    let result = [];
    let stack = [];
    if (root === null) {
        return result;
    }

    stack.push(root);
    while (stack.length !== 0) {
        let cur = stack.pop();
        result.unshift(cur.val);

        if (cur.left !== null) {
            stack.push(cur.left);
        }

        if (cur.right !== null) {
            stack.push(cur.right);
        }
    }

    return result;
};
